docs/contributors/coding-guidelines.md

```diff
@@ -201,6 +201,246 @@ alert( 'My name is ' + name + '.' );
+
+A custom type should include a description, and should always include its base type.
+
+Custom types should be named as succinctly as possible, while still retaining clarity of meaning and avoiding conflict with other global or scoped types. A `WP` prefix should be applied to all custom types. Avoid superfluous or redundant prefixes and suffixes (for example, a `Type` suffix, or `Custom` prefix). Custom types are not global by default, so a custom type does not need to be excessively specific to a particular package. However, they should be named with enough specificity to avoid ambiguity or name collisions when brought into the same scope as another type.
+
+```js
+/**
+ * A block selection object.
+ *
+ * @typedef {Object} WPBlockSelection
+ *
+ * @property {string} clientId     A block client ID.
+ * @property {string} attributeKey A block attribute key.
+ * @property {number} offset       An attribute value offset, based on the rich
+ *                                 text value.
+ */
+```
+
+Custom types can also be beneficial when describing a set of predefined options. While the [type union](https://jsdoc.app/tags-type.html) can be used with literal values as an inline type, using it in a custom type can afford the opportunity to describe the purpose of these options, and avoids a situation where aligning tags while still respecting a maximum line length of 80 characters becomes challenging due to the space occupied by a literal values union.
+
+```js
+/**
+ * Named breakpoint sizes.
+ * 
+ * @typedef {"huge"|"wide"|"large"|"medium"|"small"|"mobile"} WPBreakpoint
+ */
+```
+
+### Importing and Exporting Types
+
+Use the [TypeScript `import` function](https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html#import-types) to import type declarations from other files or third-party dependencies.
+
+Since an imported type declaration can occupy an excess of the available line length and become verbose when referenced multiple times, you are encouraged to create an alias of the external type using a `@typedef` declaration at the top of the file, immediately following [the `import` groupings](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#imports).
+
+```js
+/** @typedef {import('@wordpress/data').WPDataRegistry} WPDataRegistry */
+```
+
+Note that all custom types defined in another file can be imported.
+
+When considering which types should be made available from a WordPress package, the `@typedef` statements in the package's entrypoint script should be treated as effectively the same as its public API. It is important to be aware of this, both to avoid unintentionally exposing internal types on the public interface, and as the means by which to expose the public types of a project.
+
+```js
+// packages/data/src/index.js
+
+/** @typedef {import('./registry').WPDataRegistry} WPDataRegistry */
```

Comment:
Did you confirm it works seamlessly with VSC? I had some issues with it but I also installed VSC for the first time just before testing it 😛

Reply:
To make this work when referencing the types externally, it will require some extra changes, yes.

See also: #18927

As seen in the example there, it will start to work once that's in place.